R/functions-OnDiskMSnExp.R

```diff
@@ -481,3 +481,201 @@ precursorValue_OnDiskMSnExp <- function(object, column) {
+#'
+#' @param return.type either \code{"list"} or \code{"matrix"} to return the
+#'     result as a list or as a matrix.
+#'
+#' @param missingValue value to be used as intensity if no signal was measured
+#'     for a given rt.
+#'
+#' @param msLevel \code{integer(1)} ensuring that the chromatogram is extracted
+#'     only for a specified MS level.
+#' 
+#' @return A \code{list} or \code{matrix} with the \code{Chromatogram} objects.
+#'     If no data was present for the specified \code{rtrange} and
+#'     \code{mzrange} the function returns a \code{list} of length \code{0}.
+#'     The \code{list} is arranged first by ranges and then by files, such that
+#'     \code{result[[1]]} returns a \code{list} of \code{Chromatogram} objects
+#'     for the same rt/mz range.
+#'     For \code{return.type = "matrix"} a \code{matrix} is returned with rows
+#'     corresponding to ranges and columns to files/samples. \code{result[, 1]}
+#'     will thus return a \code{list} of \code{Chromatogram} objects for the
+#'     first sample/file, while \code{result[1, ]} returns a \code{list} of
+#'     \code{Chromatogram} objects for the same rt/mz range for all files.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.extractMultipleChromatograms <- function(x, rt, mz, aggregationFun = "sum",
+                                          BPPARAM = bpparam(),
+                                          return.type = c("list", "matrix"),
+                                          missingValue = NA_real_,
+                                          msLevel = 1L) {
+    return.type <- match.arg(return.type)
+    missingValue <- as.numeric(missingValue)
+    if (!any(.SUPPORTED_AGG_FUN_CHROM == aggregationFun))
+        stop("'aggregationFun' should be one of ",
+             paste0("'", .SUPPORTED_AGG_FUN_CHROM, "'", collapse = ", "))
+    ## Ensure we're working on one MS level only!
+    x <- filterMsLevel(x, msLevel)
+    if (length(x) == 0)
+        return(list())
+    if (missing(rt))
+        rt <- matrix(c(-Inf, Inf), nrow = 1)
+    if (missing(mz))
+        mz <- matrix(c(-Inf, Inf), nrow = 1)
+    ## Calculate the range for each row in rt 
+    if (ncol(rt) != 2)
+        rt <- t(apply(rt, MARGIN = 1, range))
+    ## Replicate if nrow rt is 1 to match nrow of mz.
+    if (nrow(rt) == 1)
+        rt <- matrix(rep(rt, nrow(mz)), ncol = 2, byrow = TRUE)
+    if (ncol(mz) != 2)
+        mz <- t(apply(mz, MARGIN = 1, range))
+    if (nrow(mz) == 1)
+        mz <- matrix(rep(mz, nrow(rt)), ncol = 2, byrow = TRUE)
+    if (nrow(rt) != nrow(mz))
+        stop("dimensions of 'rt' and 'mz' have to match")
+    ## Identify indices of all spectra that are within the rt ranges.
+    rtimes <- rtime(x)
+
+    ## 1) Subset x keeping all spectra that fall into any of the provided rt
+    ##    ranges.
+    keep_idx <- unlist(apply(rt, MARGIN = 1, function(z)
+        which(rtimes >= z[1] & rtimes <= z[2])), use.names = FALSE)
```

Comment:
I assume `rtimes` is sorted. Is there a guarantee that `rt` is sorted (meaning `rt[, 1] < rt[,2]`?):
If that is true we could reduce the comparisons and loops runs by removing overlapping `rt` ranges before (here and below in `bpmapply`). Depending on `nrow(rt)` this could be faster. Not sure if needed, e.g.:
```r
rt <- matrix(c(50, 100, 300, 400, 350, 500, 700, 900), ncol=2, byrow=TRUE)
#      [,1] [,2]
# [1,]   50  100
# [2,]  300  400
# [3,]  350  500
# [4,]  700  900
overlap <- rt[-1 ,1] <= rt[-nrow(rt), 2]
# [1] FALSE  TRUE FALSE
cbind(rt[c(TRUE, !overlap), 1L], rt[c(!overlap, TRUE), 2L])
#      [,1] [,2]
# [1,]   50  100
# [2,]  300  500
# [3,]  700  900
```
(Maybe as a new `utils....` function with some more test cases (at least some edge cases)?)